### DISCUSSION TITLE
hands-on-code/hands-on-lora

### DISCUSSION THREAD

**Discussion by giscus** [bbruceyuan/bbruceyuan.github.io#19] [Announcements]
# hands-on-code/hands-on-lora

用 PyTorch 实现从零实现 LoRA, 理解 LoRA 的原理，主要是为了展示一个 LoRA 实现的细节

https://yuanchaofa.com/hands-on-code/hands-on-lora

**Comment by ywlq** [comment]
Q1：merge=True的时候是不是需要在forward里面也判断一下，否则会不会重复计算了，比如linear+=ab，然后forward里面计算输出的时候又加上了ab*scale的结果
Q2：merge_weight函数里面加上的ab需要缩放吗，就是self.linear.weight.data +=self.scale *  (self.lora_a @ self.lora_b)

**Reply by bbruceyuan** [reply]
a1: 是
a2: 是

这两个地方都需要修改一下，我更新一下博客内容

**Reply by bbruceyuan** [reply]
已经更新，感谢提醒

**Reply by sinaitic** [reply]
你好，想问问有无merge在实际开发中的意义是什么，个人感觉就是俩种写法，就像是(W+AB)X以及WX+ABX的区别。另外关于Lora核心原理的小错误，h=W0X+ABX=W0+a/r*ABX，这里是不是有问题，应该是h=W0X+a/r*ABX=(W0+a/r*AB)X这样子

**Reply by bbruceyuan** [reply]
A1: 是这种区别，为什么要写这个是因为如果是 没有 merge 那么你就要保存两部分参数，原模型参数，LoRA 参数， Merge 之后就只需要保存一份参数，更新后的 模型参数。这也是 LoRA 重要的好处之一，推理的时候相对于原模型没有额外开销。

A2: 是有一点小问题，这里我直接把原文给揉在一起了（我那个公式3最好拆成两个部分），如果只看我这个公式就会有点蒙。已经更新，感谢提示~

**Comment by fly361** [comment]
挺好的教程， 希望能一直做下去

**Comment by AspireKuan** [comment]
fa哥，我想问一下。可不可以将lora_a和lora_b的维度换一下，变成下面的样子：
self.lora_a = nn.Parameter(torch.zeros(in_features, rank))

**Reply by AspireKuan** [reply]
```python
self.lora_b = nn.Parameter(torch.zeros(rank, out_features))
output = self.linear(X) + self.scale * ( X @ (self.lora_a @ self.lora_b))
self.linear.weight.data += self.scale * (self.lora_a @ self.lora_b).T
self.linear.weight.data -= self.scale * (self.lora_a @ self.lora_b).T
```
这样子是不是也可以呢？

**Reply by bbruceyuan** [reply]
我觉得不能 lora a 和 lora b 都初始化为零。但是从 lora 作者 Github issue 回复来看，a/b 初始化为零应该都可以（我的意见）。看你的描述计算是没有问题， Shape 对的上就行，但是最好不都初始化为零。

**Reply by AspireKuan** [reply]
OK，谢谢发哥！

**Comment by WYXsb** [comment]
矩阵A和B的形状是(out_features, rank)和(rank,in_features)，A @ B 的维度是(out_features,in_features)，而self.linear = nn.Linear(in_features, out_features),这俩形状不一样但也在merge函数里做了相加操作，请问这里是笔误吗？

**Reply by sinaitic** [reply]
代码中第23行写了# linear weight 的 Shape 是 (out_features, in_features)

**Reply by WYXsb** [reply]
哦哦，我搞反了，谢谢大佬

**Comment by YHSI5358** [comment]
直接把A和B写成两个linear层是不是一样的？也不用关注矩阵维度了，输入输出维度对的上就行

**Reply by bbruceyuan** [reply]
你这么做可以运行，但是你怎么解释呢？（以及你得想清楚的自己的目的，现在 LoRA 这种设计是为了减少可训练参数，做低秩分解）

**Reply by Alielie-9527** [reply]
如果你说的事将A和B替代原来的线性层，那么这样做是不可以的。可以去了解一下矩阵的秩，简单来说就是反映了权重能表达信息的量，对于m*n的权重矩阵，它的秩为是<=min(m,n)的，但是A和B的秩往往<=R, R是远小于m,n的，那么AB的秩也<=R ，那么矩阵AB能表示的信息是远小于W能表示的信息。

**Comment by forgetrelics** [comment]
测试的时候lora_layer的merge固定为False，无论merge与否都只会执行output = self.linear(X) + self.scale * ( X @ (self.lora_a @ self.lora_b).T )，测试结果没有差异原因是不是因为lora_b初始化为0呢？

**Reply by bbruceyuan** [reply]
不是的，你可以试一下不要设置 false，也是一样的；

**Comment by Alielie-9527** [comment]
写的非常好，我是初学者，但是想提出一点建议，并不是任意矩阵W0能分解为两个低秩矩阵，而是根据SVD分解，可以得知W0的信息主要集中在某几个维度（这个表述可能有些不好），所以AB实际上是对Δw的分解。权重的变化量 
ΔW并不需要那么高的秩。

**Reply by bbruceyuan** [reply]
AB 可以通过 delta W 分解得到，如 https://spaces.ac.cn/archives/10226 （这属于改进方法），但 naive version 就是学习出来的，并不要求可以分解

**Comment by lfxuan** [comment]
有点不清楚，为啥最后要加测试：
# Test weight merging/unmerging
lora_layer.merge_weight()
output_after_merge = lora_layer(x)
lora_layer.unmerge_weight()
output_after_unmerge = lora_layer(x)
是要比对什么吗？

**Reply by bbruceyuan** [reply]
这主要是为了: 告诉大家真实是怎么用的, 就是 merge weight 之后使用.
